### PR TITLE
ScalaCheck generator for core values via ScopedBuilder

### PR DESCRIPTION
**Pull request by shonfeder**
- Lays the foundations for #1379, on the basis of the `ScopedBuilder`.
- Initially prepared to support #2544
- The PBT of records turned up an error in the scope violation detection, for
  which I've added a regression test.
  
This is branched off of #2544 and will require

- [x] rebase once that's merged in,
- [x] the fix for #2551

- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [x] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      fields <- Gen.sequence[Seq[(String, TBuilderInstruction)], (String, TBuilderInstruction)](
+          fieldNames.map(n => Gen.zip(Gen.const(n), instr)))
+    } yield fields
+  }
+
+  /** Implicit arbitrary instances for [TBuilderInstruction]s */
+  object ArbTBuilderInstruction {
+    implicit val argContreteTlaType1: Arbitrary[TlaType1] = Arbitrary(GenTBuilderInstruction.contreteTlaType1)
+    implicit val arbTBuilderInstruction: Arbitrary[TBuilderInstruction] = Arbitrary(GenTBuilderInstruction.instr)
+  }
```
Comment (thpani) [inline tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala:122]: Why the separate object here?
It looks like scope control that could be done via access modifiers?

**Review by thpani — APPROVED** [review]
Cool stuff!
Interesting that we get a head start on the well-typed generators as a by-product 😃 

I have a bunch of low-level comments that are probably quickly resolved:

File: tlair/src/test/scala/at/forsyte/apalache/tla/lir/TlaType1Gen.scala [hunk 2]
```diff
@@ -18,8 +18,12 @@ trait TlaType1Gen {
       // produce an absolute value. Note that Math.abs(Integer.MIN_VALUE) = Integer.MIN_VALUE, so use max(0, abs(i)).
     } yield VarT1(i)
 
+  // Only monomorphic types for the values we support (i.e., not for reals)
+  def genSupportedPrimitiveMono: Gen[TlaType1] =
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply tlair/src/test/scala/at/forsyte/apalache/tla/lir/TlaType1Gen.scala]: This is unrelated to ScalaCheck, and means values we support in model checking. I've added some words to clarify 669a92ab079d64e9a7f04b2f5a2f122b39ce65fc

File: tlair/src/test/scala/at/forsyte/apalache/tla/lir/TlaType1Gen.scala [hunk 3]
```diff
@@ -152,3 +159,38 @@ trait TlaType1Gen {
   // this is the generator for arbitrary types, including nested ones
   def genType1: Gen[TlaType1] = genTypeTree
 }
+
+/**
+ * Generators for the case classes of TlaType1's that can be built into the principle supported _values_ of TlaEx. This
+ * means all types are concrete, while unsupported, non-first class, and deprecated types are excluded.
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply tlair/src/test/scala/at/forsyte/apalache/tla/lir/TlaType1Gen.scala:165]: That is a gnarly sentence! Hopefull d9fdd782698a315f3b3f3b8b16d01c7c51b07b1c improves matters.

File: tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * NOTE: The generated instruction will not have any shadowed names internally, but if a builder instance is
+     * supplied, shadowing is possible.
+     *
+     * @param typ
+     *   The type of term of which to generate a builder
+     * @param builder
+     *   Optionally and implicitly, the builder instance to use
+     */
+    def ofType(typ: TlaType1)(implicit builder: ScopedBuilder = new ScopedBuilder()): Gen[T] = {
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala]: The builder's state just tracks scope. So depending on the default builder amounts to each expression generated being in an independent scope. This Is often what's wanted. I'll add a note to explain. 5d77b0c3d9b061fda48a0c4234c00654c54e98d5

File: tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.typecomp._
+import org.scalacheck.Arbitrary
+import org.scalacheck.Gen
+
+/** Utilities for property based testing using well-typed terms via the typed builder. */
+object TypecompPBT {
+
+  /** Generators for [TBuilderInstruction]s */
+  object GenTBuilderInstruction {
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala:12]: I moved these out of the namespacing object.

File: tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala (hunk 1 above)
Comment (shonfeder) [reply tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala:123]: I've added a comment to clarify: 6bd5d3231261ec9994af5373c830ec03a081a1d5

**Pull request merged**